Additional copy to clipboard for request code and SRP payout

etra opened the pull request:
## Description

Changes for issue #266 

## Type of change

Added Copy to clipboard for request code and srp payout

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [] I have checked my code and corrected any misspellings

ppfeufer commented on aasrp/static/aasrp/javascript/view-requests.js:
```diff
@@ -52,6 +52,20 @@ $(document).ready(() => {
+                 */
+                render: (data, type) => {
+                    if (type === 'display') {
+                        return `<span>${data}</span><i class="aa-srp-fa-icon copy-text-fa-icon fa-regular fa-copy ms-2" data-clipboard-text="${data}" data-bs-tooltip="aa-srp" aria-label="Copy code to clipboard" data-bs-original-title="Copy code to clipboard"></i>`;
```
Why wrap `${data}` in a `span` but not include the copy icon in this wrapper?

`data-bs-original-title` is unnecessary, `title` is enough, Bootstrap does the rest.

"Copy request code […]" instead of just "Copy code […]"

Please ensure the string is translatable.

ppfeufer commented on aasrp/static/aasrp/javascript/view-requests.js:
```diff
@@ -92,7 +106,7 @@ $(document).ready(() => {
                 render: (data, type) => {
                     if (type === 'display') {
-                        return `<span class="srp-payout-tooltip"><span class="srp-payout-amount">${new Intl.NumberFormat(aaSrpSettings.locale).format(data)} ISK</span></span>`;
+                        return `<span class="srp-payout-tooltip"><span class="srp-payout-amount">${new Intl.NumberFormat(aaSrpSettings.locale).format(data)} ISK</span></span><i class="aa-srp-fa-icon copy-text-fa-icon fa-regular fa-copy ms-2" data-clipboard-text="${data}" data-bs-tooltip="aa-srp" aria-label="Copy payout to clipboard" data-bs-original-title="Copy payout to clipboard"></i>`;
```
There should be a wrapper `span` around this construct to include the icon.

`data-bs-original-title` is unnecessary, `title` is enough, Bootstrap does the rest.

"Copy payout amount […]" instead of just "Copy payout […]"

Please ensure the string is translatable.

etra replied:
@ppfeufer - thank you very much for all the feedback. Working on implementing it. Would you be kind and provide an example of translation of strings.

I as understand I should use: `aaSrpSettings.translation` - however I'm having hard time finding where these variables are set.

Thank you in advance

ppfeufer reviewed (changes requested):
Thank you for this PR, some small things to address, see comments.

etra commented:
I hope I have covered all the comments, please let me know if there is something extra left.

Thank you

ppfeufer commented on aasrp/templates/aasrp/view-requests.html:
```diff
@@ -51,6 +53,8 @@
                 changeSrpAmount: '{% url "aasrp:ajax_srp_request_change_payout" srp_link.srp_code 'SRP_REQUEST_CODE' %}'
             },
             translation: {
+                copyRequestCodeToClipboard: '{{ translateCopyRequestCodeToClipboard|escapejs }}',
+                copyPayoutAmountToClipboard: '{{ translateCopyPayoutAmountToClipboard|escapejs }}',
```
Please move these down to the rest of the non-filter translations.

etra replied:
Moved bellow filters

ppfeufer commented on aasrp/static/aasrp/javascript/view-requests.js:
```diff
@@ -92,7 +106,8 @@ $(document).ready(() => {
                 render: (data, type) => {
                     if (type === 'display') {
-                        return `<span class="srp-payout-tooltip"><span class="srp-payout-amount d-block cursor-pointer">${new Intl.NumberFormat(aaSrpSettings.locale).format(data)} ISK</span></span>`;
+                        return `<span class="srp-payout-tooltip"><span class="srp-payout-amount d-block cursor-pointer">${new Intl.NumberFormat(aaSrpSettings.locale).format(data)} ISK</span><i class="aa-srp-fa-icon copy-text-fa-icon fa-regular fa-copy ms-2" data-clipboard-text="${data}" data-bs-tooltip="aa-srp" aria-label="${aaSrpSettings.translation.copyPayoutAmountToClipboard}" title="${aaSrpSe ...
```
Have you tested this for side effects?

etra replied:
I have tested on my local version. Where is one issue which I do not know how to address and fix if you have suggestion - it would be amazing.

After you edit SRP value it is not affected in clipboard action - it would still copy 0 instead of actually entered value. I do not know how to trigger update of clipboard data after edit is done.

ppfeufer replied:
> I have tested on my local version. Where is one issue which I do not know how to address and fix if you have suggestion - it would be amazing.
> 
> After you edit SRP value it is not affected in clipboard action - it would still copy 0 instead of actually entered value. I do not know how to trigger update of clipboard data after edit is done.

You have to hook into the function that triggers the refresh after a new payout value is posted and ensure the value for the copy to clipboard icon is updated as well.
https://github.com/ppfeufer/aa-srp/blob/master/aasrp/static/aasrp/javascript/view-requests.js#L216-L246

But that is just one of the side effects i meant …
The other is, when I hover the copy to clipboard icon on a field that can have it's payout amount changed, I get both tooltips at the same time.
![image](https://github.com/user-attachments/assets/7451fc53-83d3-4cb3-825c-f753a85c042a)

etra replied:
👍 - will look into improving this. Thank you.

ppfeufer reviewed (changes requested):
Thanks for addressing my remarks.

Please see my new comments on this.
Also please ensure the lines in JS are not too long and become unreadable, as I mentioned in my previous review.
It is viable to split string that are too long up into variables …

etra commented:
> Thanks for addressing my remarks.
> 
> Please see my new comments on this. Also please ensure the lines in JS are not too long and become unreadable, as I mentioned in my previous review. It is viable to split string that are too long up into variables …

Any suggestion how better to split the line? By splitting more I'm getting "Trailing Whitespace" error - not sure what is the best approach here. Welcome on any input.

etra commented:
P.s. This is my first time contributing to open source project. Thank you for your understanding and patience :)

ppfeufer commented:
> Any suggestion how better to split the line? By splitting more I'm getting "Trailing Whitespace" error - not sure what is the best approach here. Welcome on any input.

Template strings over multiple lines are not ideal, as they don't compress well in the minified JS. So you have to work with variables here and split the string.

Like so:
```javascript
const ctcTranslation = aaSrpSettings.translation.copyPayoutAmountToClipboard;
const ctcIconClasses = 'aa-srp-fa-icon copy-text-fa-icon fa-regular fa-copy ms-2';
const ctcIcon = `<i class="${ctcIconClasses}" data-clipboard-text="${data}" data-bs-tooltip="aa-srp" aria-label="${ctcTranslation}" title="${ctcTranslation}"></i>`

return `<span>${data} ${ctcIcon}</span>`
```

This splits the string into smaller pieces that are easier to read and easier to maintain.

The "Trailing Whitespace" error simply means that a line has trailing whitspaces, which I don't allow.
It helps to make whitespaces visible in your IDE.

To highlight in your change:
![image](https://github.com/user-attachments/assets/4eed232b-11f6-4046-9420-2d73ec9e71e9)

All the lines I marked have a little grey dot at the end, that's a whitespace.

`Line with whitespace at the end  ` vs. `Line without whitespace at the end`

etra commented:
Hi,

I have fixed the issue with double tooltip.
Also added support for updating newValue after it is entered.

I also updated how i present values in render so it would not have whitespaces in begining so what uglify would work better.

Let me know if I should add more variables for some of longer strings.

thank you

ppfeufer commented on aasrp/static/aasrp/javascript/view-requests.js:
```diff
@@ -92,7 +110,10 @@ $(document).ready(() => {
+                        const ctcTitle = `title="${aaSrpSettings.translation.copyPayoutAmountToClipboard}"`;
+                        const ctcLabel = `aria-label="${aaSrpSettings.translation.copyPayoutAmountToClipboard}"`;
+                        const ctcIcon = `<i class="${ctcIconClass}" data-bs-tooltip="aa-srp" data-clipboard-text="${data}" ${ctcLabel} ${ctcTitle}></i>`;
+                        return `<span class="srp-payout-tooltip"><span class="srp-payout-amount d-block cursor-pointer">${new Intl.NumberFormat(aaSrpSettings.locale).format(data)} ISK</span></span> ${ctcIcon}`;
```
This is still missing a proper wrapper, so it can be addressed properly.

etra replied:
I do not know what is a proper her will this work?

ppfeufer replied:
That is not what I meant at all.
The whole HTML output needs to be wrapped in a `span` with a proper class, so the elements inside it can be addressed properly without needing to resort to `element.closest()` which should not be used at all.

ppfeufer commented on aasrp/static/aasrp/javascript/view-requests.js:
```diff
@@ -243,6 +264,11 @@ $(document).ready(() => {
         $('.srp-fleet-total-amount').html(`${new Intl.NumberFormat(aaSrpSettings.locale).format(totalSrpAmount)} ISK`);
+
+        //Update tooltip value by first selecting parent and then selecting the icon
+        const ctcElement = element.closest('td').find('i');
+        ctcElement
+            .attr('data-clipboard-text', newValue);
```
This is very janky and not bullet proof. `element.closest()` should not be used, no matter what GPT tells you.

etra commented on aasrp/static/aasrp/javascript/view-requests.js:
```diff
@@ -243,6 +265,13 @@ $(document).ready(() => {
+
+        //Update tooltip value by first selecting parent and then selecting the icon
+        const outerSpan = element.parent();
+        const tdElement = outerSpan.parent();
```
this is not td anymore it's a spanWrapper...

ppfeufer commented:
Superseded by #272
